api/turing/config/config.go

```diff
@@ -122,72 +145,169 @@ type RouterDefaults struct {
-	}
+// FromFiles creates a Config object from config files. JSON and YAML format are supported.
+//
+// If multiple config files are provided, the subsequent config files will override the config
+// values from the config files loaded earlier.
+//
+// These config files will override the default (refer to setDefaultValues function) config values
+// and can be overridden by the values from environment variables. Nested keys in the config
+// can be set from environment variable name separed by "_". For instance the config value for
+// "DbConfig.Port" can be overridden by environment variable name "DBCONFIG_PORT". Note that
+// all environment variable names must be upper case.
+//
+// Refer to example.yaml for an example of config file.
+func FromFiles(filepaths ...string) (*Config, error) {
+	v := viper.New()
 
-	// Run config validator
-	validate, err := newConfigValidator()
-	if err != nil {
-		return nil, fmt.Errorf("Unable to init Config validator: %v", err)
+	// Load default config values
+	setDefaultValues(v)
+
+	// Load config values from the provided config files
+	for _, f := range filepaths {
+		v.SetConfigFile(f)
+		err := v.MergeInConfig()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read config from file '%s': %s", f, err)
+		}
 	}
-	err = validate.Struct(cfg)
+
+	// Load config values from environment variables.
+	// Nested keys in the config is represented by variable name separated by '_'.
+	// For example, DbConfig.Host can be set from environment variable DBCONFIG_HOST.
+	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	v.AutomaticEnv()
+
+	config := &Config{}
+
+	// Unmarshal config values into the config object.
+	// Add StringToQuantityHookFunc() to the default DecodeHook in order to parse quantity string
+	// into quantity object. Refs:
+	// https://github.com/spf13/viper/blob/493643fd5e4b44796124c05d59ee04ba5f809e19/viper.go#L1003-L1005
+	// https://github.com/mitchellh/mapstructure/blob/9e1e4717f8567d7ead72d070d064ad17d444a67e/decode_hooks_test.go#L128
+	err := v.Unmarshal(config, func(c *mapstructure.DecoderConfig) {
+		c.DecodeHook = mapstructure.ComposeDecodeHookFunc(
+			mapstructure.StringToTimeDurationHookFunc(),
+			mapstructure.StringToSliceHookFunc(","),
+			StringToQuantityHookFunc(),
+		)
+	})
 	if err != nil {
-		return nil, fmt.Errorf("Config validation failed: %v", err)
+		return nil, fmt.Errorf("failed to unmarshal config values: %s", err)
 	}
 
-	return &cfg, nil
+	return config, nil
+}
+
+// setDefaultValues for all keys in Viper config. We need to set values for all keys so that
+// we can always use environment variables to override the config keys. In Viper v1, if the
+// keys do not have default values, and the key does not appear in the config file, it cannot
+// be overridden by environment variables, unless each key is called with BindEnv.
+// https://github.com/spf13/viper/issues/188
+// https://github.com/spf13/viper/issues/761
+func setDefaultValues(v *viper.Viper) {
```

Comment:
Just a blunt idea: since we support multiple config files, wouldn't it bu easier, clearer to have a default config file with listed default values that is implicitly loaded before any other user-passed config files?

Reply:
Yes my original intention is actually to have a default config yaml file that is always loaded. This can also serve as a documentation. 

However, if we want to have this default file that is always loaded, when we build the Turing API binary, we want to embed this file in the binary. There are numerous way to do this e.g the good old `go-bindata`. However this will require an additional step in the build process to embed the default file, before running `go build` step. I'd rather not have to modify the build step if possible.

Actually I am kinda waiting for the embed feature to be natively available in go1.16
https://github.com/golang/go/issues/41191

Another way is to have this default specified in say `default.go` like
```go
defaultConfig := `
DbConfig:
  Host: localhost
  Port: 5432
  ... and so on ..
`
```
But I guess this approach is not that much better than our current way of defining defaults with `viper.setDefaults` as both are not very explicit with users having to browse the source code to check the default.

Reply:
Is it really necessary to embed the default config file? Would it be possible to just keep it side by side with the binary, in the same way as we do with `swagger.yaml` file?

Reply:
That actually is possible because we copy the swagger.yaml explicitly in the Dockerfile
https://github.com/gojek/turing/blob/main/api/Dockerfile#L32
And we run Turing binary in the directory where we copy swagger.yaml to.

But I imagine if I only have a Turing API server binary, then it becomes an assumption that the directory where I run this Turing binary from have the respective yaml files, which I think can be a surprise for some people.

Reply:
I will merge this in for now. We can add default configuration based on file later. But I believe the default config is usually embeded in the binary itself.

https://github.com/prometheus/prometheus/blob/4eca4dffb89d9da927f1e7884efe85cbd6dc30f3/config/config.go#L71

https://github.com/k0sproject/k0s/blob/14f348158fc8aba111bb074face2338fd2a05849/pkg/apis/v1beta1/cluster.go#L114